### ISSUE TITLE
Rule types?

### ISSUE THREAD

**Issue by michelbetancourt** [benas/easy-rules#12]
It seems like different types of rules require different processors to fire them.  Is there a reason for this or am I missing something?

**Comment by benas** [comment]
Rules implementing the `Rule` interface or extending the `BasicRule` class should be registered in a `DefaultRulesEngine`. Rules which are annotated Pojos should be registered in a `AnnotatedRulesEngine`.  Is this what you mean by different types of rules?

**Comment by michelbetancourt** [comment]
Yes.

**Comment by mrcritical** [comment]
I think this question is inline with my question in #11. It does seem like `Rule` should be used in either case, ultimately, for consistency. That way the `RuleEngine<R>` can become a little easier to deal with by becoming `RuleEngine<R extends Rule>`. It will be easier to write new rule engines this way too. Just a suggestion.

Of course the listener proposed in #11 fixes my issue (assuming it works on all `RuleEngine` implementations).

**Comment by benas** [comment]
The current implementation of Easy Rules is not consistent, I agree. There should be only one `RulesEngine` implementation that would be able to process both annotated pojos and implementations of the `Rule` interface in a consistent way.

I have planned to make annotated pojos implement the `Rule` interface at runtime using dynamic proxies. This way, users can register both types of rules without having to choose the right rules engine.

I implemented a poc on how the approach will work:

```java
import org.easyrules.annotation.Action;
import org.easyrules.annotation.Condition;
import org.easyrules.api.Rule;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

public class RuleProxyTest {

    public static void main(String[] args) throws Exception {

        MyRule myRule = new MyRule();
        RuleProxy ruleProxy = new RuleProxy(myRule);
        Rule rule = (Rule) Proxy.newProxyInstance(Rule.class.getClassLoader(), new Class[]{Rule.class}, ruleProxy);
        rule.performActions();

    }

    static class RuleProxy implements InvocationHandler {

        private Object target;

        public RuleProxy(Object target) {
            this.target = target;
        }

        @Override
        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
            if (method.getName().equals("performActions")){
                Method toInvoke = findActionMethod();
                return toInvoke != null? toInvoke.invoke( target, args ) : null;
            }
            return null;
        }

        private Method findActionMethod(){
            try {
                Method[] declaredMethods = target.getClass().getDeclaredMethods();
                for (Method method : declaredMethods) {
                    if (method.isAnnotationPresent(Action.class)) {
                        return method;
                    }
                }
                return null;
            }
            catch( Exception e) {
                throw new RuntimeException(e);
            }
        }

    }

    @org.easyrules.annotation.Rule
    public static class MyRule {

        @Condition
        public boolean when() {
            return true;
        }

        @Action
        public void then() {
            System.out.println("rule executed");
        }

    }
}
```

This is planned for next release, so keep tuned ;-)

Regards
Mahmoud

**Comment by mrcritical** [comment]
Excellent. We use proxies in similar way. That would be a very nice improvement.

**Comment by benas** [comment]
Hi @michelbetancourt , @mrcritical 

I've released a new version (2.0) that provides a unified interface `RulesEngine` through which both annotated POJOs and `Rule` implementations can be used in a consistent way.

With this new improvement, there is no more need to use different types of engines for different types of rules. Everything is streamlined.

@michelbetancourt I would like to thank you because with a simple question, you made me redesign Easy Rules to improve its API :-)

Best regards,
Mahmoud

**Comment by michelbetancourt** [comment]
Glad to hear it worked out!
